### PR TITLE
fix: do not start DWR controller if devworkspaces are not enabled

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
fix: do not start DWR controller if devworkspaces are not enabled.
I'll add a bit more context soon.

### What issues does this PR fix or reference?
fixes https://github.com/eclipse/che/issues/20605

### How to test this PR?
I've built the image `sleshchenko/che-operator:dwco-idling`
The test scenario with chectl operator installer is the following:
1. Deploy Che with Che Server workspace engine (`chectl server:deploy --installer=operator --platform=openshift --che-operator-image=sleshchenko/che-operator:dwco-idling`);
Expected: no CRDs are available, no DWO controller installed, DWCO is not activated (can be checked by logs)

```
2021-10-08T09:26:25.731Z	INFO	setup	DevWorkspace support disabled. Will initiate restart when CheCluster with devworkspaces enabled will appear
```

2. Modify existing CheCluster and enable DevWorkspace.
Expected: Che operator detects that change and reboot(after delay up to 1minute), install DWO, reboot, activate DWCO;

3. Modify existing CheCluster and disable DevWorkspace, restart Che Operator.
Expected: CRD is present but no CheCluster with devworkspaces exists, DWCO is not activated.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [ ] [OLM bundles are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -297,54 +298,78 @@ func enableDevworkspaceSupport(mgr manager.Manager) error {
 		return err
 	}
 
-	// we only enable Devworkspace support, if there is the controller.devfile.io resource group in the cluster
-	// we assume that if the group is there, then we have all the expected CRs there, too.
+	// we only enable DevWorkspace support, if there is the controller.devfile.io resource group in the cluster
+	// (we assume that if the group is there, then we have all the expected CRs there, too.)
+	// and DevWorkspaces are enabled at least on one CheCluster
```
Comment (metlos) [inline main.go:303]: This would mean that we'd have to restart the operator at each "flip" of the `Spec.DevWorkspace.Enable`.

File: main.go [hunk 2]
```diff
@@ -275,18 +279,17 @@ func main() {
...
 	}
 }
 
-func enableDevworkspaceSupport(mgr manager.Manager) error {
+func enableDevworkspaceSupport(mgr manager.Manager, stop <-chan struct{}) error {
 	// DWO and DWCO use the infrastructure package for openshift detection. It needs to be initialized
 	// but only supports OpenShift v4 or Kubernetes.
 	if err := infrastructure.Initialize(); err != nil {
-		setupLog.Info("devworkspace cannot run on this infrastructure")
+		setupLog.Error(err, "failed to evaluated infrastructure which is needed for DevWorkspace support")
```
Comment (amisevsk) [inline main.go:292]:
```suggestion
		setupLog.Error(err, "failed to evaluate infrastructure which is needed for DevWorkspace support")
```

File: main.go [hunk 3]
```diff
@@ -297,54 +303,107 @@ func enableDevworkspaceSupport(mgr manager.Manager) error {
...
+	}
+
+	setupLog.Info("DevWorkspace support enabled")
+
+	return nil
+}
+
+func doesCheClusterWithDevWorkspaceEnabledExist(mgr manager.Manager) (bool, error) {
+	cheClusters := &orgv1.CheClusterList{}
+	err := mgr.GetClient().List(context.TODO(), cheClusters, &client.ListOptions{})
```
Comment (sleshchenko) [inline main.go:358]: it does not work at this stage.
```
2021-10-07T18:12:24.344+0300	ERROR	setup	unable to initialize devworkspace support	{"error": "the cache is not started, can not read objects"}
main.main
```
@tolusha @metlos any ideas how to do that check at this point?

Alternative - we can drop backward compatibility fallback https://github.com/eclipse-che/che-operator/blob/main/pkg/deploy/dev-workspace/dev_workspace.go#L88
and just check DevWorkspace allowed env var, as @metlos proposed.

**Comment by sleshchenko** [comment]
devworkspace happy path test fails with:
```
2021-10-07T16:05:32.127Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": ":60000"}
panic: Attempting to determine information about the cluster without initializing first

goroutine 1 [running]:
github.com/devfile/devworkspace-operator/pkg/infrastructure.IsOpenShift(...)
	/che-operator/vendor/github.com/devfile/devworkspace-operator/pkg/infrastructure/cluster.go:65
github.com/eclipse-che/che-operator/controllers/devworkspace.(*CheClusterReconciler).SetupWithManager(0xc00021d5f0, 0x2015400, 0xc00000a1e0, 0x0, 0x0)
	/che-operator/controllers/devworkspace/controller.go:124 +0x49b
main.main()
	/che-operator/main.go:268 +0x3f0
```

File: main.go (hunk 3 above)
Comment (sleshchenko) [reply main.go]: Starting manager is a blocking operation, so at start up point we don't have better alternative than using non-cached client.

File: controllers/devworkspace/setup/setup.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+//   Red Hat, Inc. - initial API and implementation
+//
+package setup
+
+import (
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+func doAsyncRestartWhenDevWorkspaceEnabled(mgr manager.Manager, stop <-chan struct{}) {
+
```
Comment (metlos) [inline controllers/devworkspace/setup/setup.go:19]: Is this meant to be empty?

File: controllers/devworkspace/state.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+const (
+	APINotPresentState DevWorkspaceState = 0
+	DisabledState      DevWorkspaceState = 1
+	EnabledState       DevWorkspaceState = 2
+)
+
+// ShouldDevWorkspacesBeEnabled evaluates if DevWorkspace mode should be enabled
+// which we only do, if there is the controller.devfile.io resource group in the cluster
+// and DevWorkspaces are enabled at least on one CheCluster
+func ShouldDevWorkspacesBeEnabled(mgr manager.Manager) (bool, error) {
```
Comment (tolusha) [inline controllers/devworkspace/state.go:38]: I would say it is more about `can` than `should`

File: controllers/devworkspace/state.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+			Scheme: mgr.GetScheme(),
+		})
+		if err != nil {
+			return false, err
+		}
+		nonCachedClient = &c
+	}
+
+	cheClusters := &orgv1.CheClusterList{}
+	err := (*nonCachedClient).List(context.TODO(), cheClusters, &client.ListOptions{})
```
Comment (tolusha) [inline controllers/devworkspace/state.go:107]: Should we find only in a watch namespace (if set)

File: controllers/devworkspace/state.go (hunk 5 above)
Comment (sleshchenko) [reply controllers/devworkspace/state.go:38]: that's can and should )

Comment (sleshchenko) [reply controllers/devworkspace/state.go:38]: We should not care about multiple operators with devworkspace enabled. Should we?

**Review by metlos — APPROVED** [review]
Manually tests on minikube. Seems to be working as expected 🎉

File: controllers/devworkspace/state.go (hunk 6 above)
Comment (sleshchenko) [reply controllers/devworkspace/state.go:107]: OK, we got an agreement that it's OK enough

**Review by ibuziuk — APPROVED** [review]
verified via cluster-bot
```2021-10-08T15:10:52.278Z	INFO	setup	DevWorkspace support disabled. Will initiate restart when CheCluster with devworkspaces enabled will appear```

 :+1:

**Pull request merged**